katip-elasticsearch/src/Katip/Scribes/ElasticSearch/Internal.hs

```diff
@@ -518,3 +510,156 @@ instance ESVersion ESV5 where
+                 , Functor (BH v IO)
+#endif
+                 )
+    => EsScribeCfg v
+    -> BHEnv v
+    -> IndexName v
+    -- ^ Treated as a prefix if index sharding is enabled
+    -> MappingName v
+    -> Severity
+    -> Verbosity
+    -> IO Scribe
+mkEsBulkScribe cfg@EsScribeCfg {..} env ix mapping sev verb = do
+  (inChan, outChan) <- UNB.newChan $ unEsQueueSize essQueueSize
+  endSig <- newEmptyMVar
+  workerSig <- newEmptyMVar
+
+  runBH prx env $ do
+    chk <- indexExists prx ix
+    -- note that this doesn't update settings. That's not available
+    -- through the Bloodhound API yet
+    unless chk $ void $ do
+      r1 <- createIndex prx essIndexSettings ix
+      unless (statusIsSuccessful (responseStatus r1)) $
+        liftIO $ throwIO (CouldNotCreateIndex r1)
+      r2 <- if shardingEnabled
+              then putTemplate prx tpl tplName
+              else putMapping prx ix mapping base
+      unless (statusIsSuccessful (responseStatus r2)) $
+        liftIO $ throwIO (CouldNotCreateMapping r2)
+
+  workers <- replicateM (unEsPoolSize essPoolSize) $ async $
+    startBulkWorker cfg env mapping workerSig outChan
+
+  _ <- async $ do
+    takeMVar endSig
+    putMVar workerSig ()
+    -- atomically $ closeTBMQueue q
+    mapM_ waitCatch workers
+    putMVar endSig ()
+
+  let finalizer = putMVar endSig () >> takeMVar endSig
+  return (Scribe (logger inChan) finalizer)
+  where
+    logger :: forall a
+            . LogItem a
+           => UNB.InChan (IndexName v, Value)
+           -> Item a
+           -> IO ()
+    logger inChan i = when (_itemSeverity i >= sev) $
+      UNB.writeChan inChan (chooseIxn prx ix essIndexSharding i, itemJson' i)
+
+    prx :: Typeable.Proxy v
+    prx = Typeable.Proxy
+
+    tplName = toTemplateName prx ixn
+
+    shardingEnabled = case essIndexSharding of
+      NoIndexSharding -> False
+      _               -> True
+
+    tpl = toIndexTemplate prx (toTemplatePattern prx (ixn <> "-*")) (Just essIndexSettings) [toJSON base]
+
+    base = baseMapping prx mapping
+
+    ixn = fromIndexName prx ix
+
+    itemJson' :: LogItem a => Item a -> Value
+    itemJson' i
+      | essAnnotateTypes = itemJson verb (TypeAnnotated <$> i)
+      | otherwise        = itemJson verb i
+
+
+data DiscrimLen =
+    SendIt
+  | StackIt
+  deriving Show
+
+startBulkWorker
+    :: forall v . (ESVersion v, V5.MonadBH (BH v IO))
+    => EsScribeCfg v
+    -> BHEnv v
+    -> MappingName v
+    -> MVar ()
+    -> UNB.OutChan (IndexName v, Value)
+    -> IO ()
+startBulkWorker _ env mapping dieSignal outChan = do
+  -- We need to randomize upload delay
+  -- so that workers do not stampede the nodes
+  delayTime' <- randomRIO (100, 2000)
+  go (delayTime' * milliseconds) [] 0
+  where
+    milliseconds :: Int
+    milliseconds = 1000
+
+    go :: Int -> [V5.BulkOperation] -> Int -> IO ()
+    go delayTime xs len = do
+      result <- race (UNB.readChan outChan) (threadDelay delayTime)
```

Comment:
I'm a little bit worried about the safety of this race. IIRC race will run 2 async threads and will cancel the losing thread after the winner returns. Is it possible for the delay to hit simultaneously with an item being removed from the queue effectfully, the thread to be canceled and that item to be dropped on the floor?

It seems like maybe [readChanOnException](https://hackage.haskell.org/package/unagi-chan-0.4.0.0/docs/Control-Concurrent-Chan-Unagi-Bounded.html#v:readChanOnException) could be used to make this async exception safe (and the ThreadKilled exception would be delivered asynchronously). I guess the danger there is if the chan is full and the `readChan` was blocking, in the cleanup function we provide to `readChanOnException` you'd have to do a `tryWriteChan` to make a best effort to give it back without blocking.

Reply:
I talked to Brandon Simmons, he said to spin tryRead in a loop for this purpose.

Reply:
Suppose that makes sense. I'm kind of curious how hard that would be on the CPU. We're not in STM so we don't have a lot of options that wake up your loop only when things change. Does he suggest a check delay or anything to mitigate that?

Reply:
Exact wording, hopefully he won't mind: "So if you've got a single reader then the clearest way to do what you want would probably be to use `tryReadChan` and make sure that if a blocking read times out that you retry (that's what tryReadChan gives you: the ability to block, or busy-wait and retry as much as you like), before proceeding with another readChan on the queue."

Sounds like busy-spin as we discussed. It's not clear to me how this avoids a faulty read if the raced timeout lands in the middle of the `readChan`.

Some arithmetic could be used to wait rather than attempting the genuine read if the time differential is within #{X} nanos of the timeout deadline but I don't like hitting up the clock in a hot loop like this.

Reply:
Thanks for the additional clarification. Sounds good. I say let's move forward with this simple fallback-to-spin-loop approach and see where it gets us.

Reply:
https://github.com/bitemyapp/break-unagi-chan

Seeing what breaks and what works here.

Reply:
https://github.com/bitemyapp/break-unagi-chan Here's some results from my attempt at shaking out how unagi-chan would behave in the pessimal case.

I'm not 100% certain how to interpret the results (love to know what you think), but I got basically the same results from the regular and "exception safe" variants.

As it stands, I'd say that I would personally be comfortable going forward with this variant as a harm reducing relative to the current scribe's behavior (the load-shedding can get extreme)

All my alternatives would involve load-shedding (STM of bounded collection), livelock (TQueue, TBQueue), or unbounded memory use (STM of unbounded collection). I'd also have to be careful to not make the same mistake, whatever it was, that induces livelock in TQueue/TBQueue/TBMQueue.

I'll focus on exposing configuration bits for the moment, I'd like to know what you think meanwhile.

Reply:
Thanks for the research! When you say going forward with this variant, you mean with the tryReadChan as suggested by the unagi-chan author?

Reply:
Incoming text wall.

Took a closer look at the code and I'm starting to be convinced this is pretty hard to test. In your code you fill a chan with all the data that will ever be pushed to it. You then try your algorithms and give them decreasing "patience" for how long they'll wait to read data that we're certain is in the chan, down to like 1 microsecond (if I understand correctly) which I think may even be below the practicality of the scheduler. Eventually I would expect this to break down but I think this may just be testing the constant factors of each mechanism and not the safety.

The hypothetical data loss scenario we're talking about is when data is available to the chan at almost exactly the same time as the worker will give up, with the idea that a read from the chan isn't atomic and could get interrupted by a timeout and drop the message on the floor. In order to compare this across implementations somewhat deterministically I think you'd start with an empty queue and then coordinate many pushes to it that are all right around the time where it'd give up (and then have your implementation retry in any case). Like running a particle accelerator and looking for collisions. I think that means your implementation under test would need to keep track of the last number it saw so it could simply abort on first anomaly. That way, it could differentiate between successfully timing out but getting the data the next time and actually losing data.

So it seems like from the code there's 3 variations we want to consider:
1. race blocking read with threadDelay
2. race blocking read w/ attempted re-queueing with thread delay (so, theoretically safer, maybe bad perf, maybe not effective at preventing lost writes)
3. blocking version of tryReadChan.

I think we can probably eliminate number 3. It seems like in the unagi code, that's exactly equivalent to 1:

```haskell
tryReadChan :: OutChan a -> IO (UT.Element a, IO a)
tryReadChan oc = do -- no mask necessary
    (seg,segIx) <- startReadChan oc
    return ( ... ellided, promise verison unused... , readSegIxUnmasked id (seg,segIx))

readChan :: OutChan a -> IO a
readChan = \oc-> startReadChan oc >>= readSegIxUnmasked id
```

So now we're down to: is `readChanOnException` with timeout safer than `readChan` with timeout? If so, is there a performance impact to using `readChanOnException`?

For checking safety, I'm not sure the best way but maybe we should just batter a queue with data and abort on first anomaly. So maybe:
1. Spawn a writer with random short publish delays roughly around the timeout in the reader.
2. Spawn a reader that retries, and remembers the last number it saw and blows up on first out of sequence.
3. Run the program for a while and see if we can *ever* produce the bug. We would do this with plain `readChan` first.

I feel like if we can get up to some high number of attempts without creating this bug, then we can probably just stop worrying about it, race `readChan` and call it a day. If we do create the bug, see if we can recreate it using the "safe" `readChan`. If not, see if perf disqualifies the safe variant.

I'll take a crack locally at your repo to see if I can rig this test setup.